internal/verify/test_fixture.go

```diff
@@ -0,0 +1,203 @@
+	}
+	tf.resources, err = loadResources(fsys, path)
+	if err != nil {
+		return nil, err
+	}
+	return tf, nil
+}
+
+func loadResources(fsys fs.FS, path string) (map[string]*v1.Resource, error) {
+	pb := &policyv1.TestFixture_Resources{}
+	file, err := util.OpenOneOfSupportedFiles(fsys, filepath.Join(path, ResourcesFileName))
+	if err != nil || file == nil {
+		return nil, err
+	}
+	defer file.Close()
+	err = util.ReadJSONOrYAML(file, pb)
+	if err != nil {
+		return nil, err
+	}
+	return pb.Resources, nil
+}
+
+func loadPrincipals(fsys fs.FS, path string) (map[string]*v1.Principal, error) {
+	pb := &policyv1.TestFixture_Principals{}
+	file, err := util.OpenOneOfSupportedFiles(fsys, filepath.Join(path, PrincipalsFileName))
+	if err != nil || file == nil {
+		return nil, err
+	}
+	defer file.Close()
+	err = util.ReadJSONOrYAML(file, pb)
+	if err != nil {
+		return nil, err
+	}
+	return pb.Principals, nil
+}
+
+func (tf *testFixture) runTestSuite(ctx context.Context, eng *engine.Engine, shouldRun func(string) bool, file string, ts *policyv1.TestSuite) (SuiteResult, bool) {
+	failed := false
+
+	sr := SuiteResult{File: file, Suite: ts.Name}
+	if ts.Skip || !shouldRun(ts.Name) {
+		sr.Skipped = true
+		return sr, failed
+	}
+	tests, err := tf.getTests(ts)
+	if err != nil {
+		failed = true
+		sr.Tests = []TestResult{{
+			Name:    "Failed to load the test suite",
+			Skipped: false,
+			Failed:  true,
+			Error:   err.Error(),
+		}}
+		return sr, failed
+	}
+	for _, test := range tests {
+		if err := ctx.Err(); err != nil {
+			return sr, failed
+		}
+
+		testResult := TestResult{Name: test.Name}
+		if test.Skip || !shouldRun(test.Name) {
+			testResult.Skipped = true
+			sr.Tests = append(sr.Tests, testResult)
+			continue
+		}
+
+		actual, err := eng.Check(ctx, []*v1.CheckInput{test.Input})
+		if err != nil {
+			testResult.Failed = true
+			testResult.Error = err.Error()
+			failed = true
+			sr.Tests = append(sr.Tests, testResult)
+			continue
+		}
+
+		if len(actual) == 0 {
+			testResult.Failed = true
+			testResult.Error = "Empty response from server"
+			failed = true
+			sr.Tests = append(sr.Tests, testResult)
+			continue
+		}
+
+		// EffectsMatch is a type created to make the diff output nicer.
+		type EffectsMatch map[string]effectv1.Effect
+		expectedResult := EffectsMatch(test.Expected)
+
+		actualResult := make(EffectsMatch, len(actual[0].Actions))
+		for key, actionEffect := range actual[0].Actions {
+			actualResult[key] = actionEffect.Effect
+		}
+
+		if diff := cmp.Diff(expectedResult, actualResult); diff != "" {
+			testResult.Failed = true
+			testResult.Error = diff
+			failed = true
+		}
+
+		sr.Tests = append(sr.Tests, testResult)
+	}
+
+	return sr, failed
+}
+
+var (
+	ErrPrincipalNotFound = errors.New("principal not found")
+	ErrResourceNotFound  = errors.New("resource not found")
+)
+
+func formatTestName(tableTestName, principal string) string {
+	return fmt.Sprintf("%q by principal %q", tableTestName, principal)
```

Comment:
Just wondering aloud here: should we make this structured (machine and human readable)? Is there a benefit to that?

The normal output looks OK

```
Test results
= TestSuite (suite_test.yaml) 
== "John and his leave request" by principal "john" [OK]
```

But, the JSON output not so much:

```
go run ./cmd/cerbos compile --tests=internal/test/testdata/verify internal/test/testdata/store -f json
{
  "results": [
    {
      "file": "suite_test.yaml",
      "suite": "TestSuite",
      "tests": [
        {
          "name": "\"John and his leave request\" by principal \"john\""
        }
      ]
    }
  ]
}
```

Reply:
What about wrapping these string values in single quotes?

Reply:
Also realised that the JSON output doesn't show whether the test passed or failed -- which is a bug that has existed even before this refactoring. Mea culpa.

Reply:
> What about wrapping these string values in single quotes?

How about converting `Test.name` field from a `string` to a `message` that can hold both the test name and the principal name?. Then the output function can determine how to format it based on context.

Reply:
Fixed both above issues.

Reply:
Having field `name` of type struct is a bit unintuitive. What do you think about adding principal to TestResult type instead and make TestResult implement the `Stringer` interface to produce the value we have now.

Reply:
>Also realised that the JSON output doesn't show whether the test passed or failed -- which is a bug that has existed even before this refactoring. Mea culpa.

Fixed.

Reply:
> Having field `name` of type struct sounds a bit unintuitive to me in this context. What do you think about adding principal to TestResult type instead and make TestResult implement the `Stringer` interface to produce the value we have now.

Sounds good to me.

Reply:
>How about converting Test.name field from a string to a message that can hold both the test name and the principal name? Then the output function can determine how to format it based on context.

I ended up going this way. Please have a look.
```
go run ./cmd/cerbos compile --tests=internal/test/testdata/verify internal/test/testdata/store -f json
{
  "results": [
    {
      "file": "suite_test.yaml",
      "suite": "TestSuite",
      "tests": [
        {
          "name": {
            "name": "John and his leave request",
            "principal": "john"
          },
          "failed": false
        }
      ]
    }
  ]
}
```
`name.name` doesn't look good. Probably need to rename the inner `name` to something else.

Reply:
How about calling the outer field `case`? Then it would look like the following;

```json
{
  "results": [
    {
      "file": "suite_test.yaml",
      "suite": "TestSuite",
      "tests": [
        {
          "case": {
            "name": "John and his leave request",
            "principal": "john"
          },
          "failed": false
        }
      ]
    }
  ]
}
```

Reply:
Done